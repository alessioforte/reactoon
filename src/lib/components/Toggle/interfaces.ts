import React from 'react'

export interface Data {
  checked: boolean,
  label?: string
}

export interface Props {
  label?: string,
  initialValue: boolean,
  theme: any,
  onChange: (e: React.ChangeEvent<HTMLDivElement>, data: Data) => void
}

export interface KnobProps {
  active: boolean
}
