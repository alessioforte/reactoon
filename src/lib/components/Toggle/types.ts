import React from 'react'

export type Data = {
  checked: boolean,
  label?: string,
  name?: string,
  value: boolean
}

export type Props = {
  label?: string,
  name?: string,
  initialValue?: boolean,
  onChange?: (data: Data, e: React.ChangeEvent<HTMLDivElement>) => void
  theme: any,
}

export type KnobProps = {
  active: boolean,
  name?: string,
  value: boolean
}
