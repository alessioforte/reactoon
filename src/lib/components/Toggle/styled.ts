import styled from 'styled-components'
import { KnobProps } from './interfaces'
import { focus } from '../Styled/css'

export const Block = styled.div`
  display: inline-block;
`
export const Knob = styled.div<KnobProps>`
  flex-shrink: 0;
  position: relative;
  height: 20px;
  width: 40px;
  border-radius: 20px;
  position: relative;
  background: ${props => props.theme.colors.background};
  cursor: pointer;
  transition: all 0.3s ease;
  &:hover {
    opacity: 0.8;
  }
  &:focus {
    ${focus}
  }
  &:after {
    content: '';
    height: 16px;
    width: 16px;
    background: ${props =>
      props.active ? props.theme.colors.primary : props.theme.colors.ground};
    left: ${props => props.active ? '20px' : '0'};
    border-radius: 50%;
    position: absolute;
    margin: 2px;
    transition: all 0.3s ease;
  }
`;
export const Label = styled.div`
  display: inline-block;
  margin-left: 5px;
`;
