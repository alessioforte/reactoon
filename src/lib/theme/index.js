/**
 * Default Theme
 */

const colors = {
    primary: '#ffff00',
    secondary: '#DAA520',
    action: '',
    success: '#00FA9A',
    warning: '#FFA500',
    danger: '#B22222',
    error: '#B22222',
    info: '#00BFFF',
    light: '#fff',
    dark: '#1a1a1a',
    text: '#fff',
    textLight: '#ffffff',
    textDark: '#000000',
    textIdle: '',
    textActive: '',
    textError: '',
    groundzero: '#EFEFEF',
    background: '#eee',
    flatground: '#D3D3D3',
    ground: '#B5B5B5',
    hillground: '#848484',
    highground: '#3A3A3A',
    active: '#FFD700',
    selected: '#999',
    hover: '#2182BD',
    idle: '#888',
    disabled: '#778899',
    focus: 'none',
    holiday: '#FF4365',
    shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
}

const borders = {
    radius: 0
}

export const styles = {
    colors,
    borders
}

export function getContrastYIQ(hexcolor) {
    if (typeof hexcolor !== 'string') throw new Error(`${hexcolor} is not a string`)
    if (hexcolor[0] !== '#') throw new Error(`${hexcolor} - HEX color not valid`)

    hexcolor = hexcolor.replace('#', '');

    if (hexcolor.length < 3) throw new Error(`${hexcolor} - HEX color not valid`)
    if (hexcolor.length === 3) hexcolor = hexcolor + hexcolor

    var r = parseInt(hexcolor.substr(0,2),16);
    var g = parseInt(hexcolor.substr(2,2),16);
    var b = parseInt(hexcolor.substr(4,2),16);

    var yiq = ((r*299)+(g*587)+(b*114))/1000;

    return (yiq >= 128) ? 'dark' : 'light';
}

export default {
    styles,
    getContrastYIQ
}
