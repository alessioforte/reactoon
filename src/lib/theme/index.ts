/**
 * Default Theme
 */

const COLORS = {
  primary: '#ffff00',
  secondary: '#DAA520',
  success: '#00FA9A',
  warning: '#FFA500',
  danger: '#B22222',
  error: '#B22222',
  info: '#00BFFF',
  light: '#fff',
  dark: '#1a1a1a',
  // text: '#fff',
  // textLight: '#ffffff',
  // textDark: '#000000',
  // textIdle: '',
  // textActive: '',
  // textError: '',

  groundzero: '#EFEFEF',
  background: '#eee',
  flatground: '#D3D3D3',
  ground: '#B5B5B5',
  upperground: '#848484',
  higherground: '#3A3A3A',

  active: '#FFD700',
  selected: '#999',
  hover: '#2182BD',
  idle: '#888',
  disabled: '#778899',
  focus: '#61dafb',
  holiday: '#FF4365',
  shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
};

const BORDER = {
  radius: 0,
  width: '1px'
};

const SIZE = {
  is: 'xs',
  xs: {
    height: '',
    font: ''
  },
  sm: {},
  md: {},
  lg: {},
  xl: {}
};

export const styles = {
  colors: COLORS,
  border: BORDER,
  size: SIZE
};

export function create(theme) {
  let colors = COLORS;
  let border = BORDER;
  if (theme.colors) colors = { ...colors, ...theme.colors };
  if (theme.border) border = { ...border, ...theme.border };
  return {
    colors,
    border
  };
}

export function getContrastYIQ(hex): string {
  const color = validateHex(hex);
  if (!color) {
    console.warn(`Something goes wrong getContrastYIQ not applied`);
    return 'dark';
  }
  const [r, g, b] = colorHexToRgb(color);
  const yiq = (r * 299 + g * 587 + b * 114) / 1000;
  return yiq >= 128 ? 'dark' : 'light';
}

export function getColorLuminance(hex, luminance): string {
  const color = validateHex(hex);
  if (!color) {
    console.warn(`Something goes wrong getColorLuminance not applied`);
    return hex;
  }
  const rgb = colorHexToRgb(color);
  hex = '#';
  luminance = luminance || 0;
  rgb.forEach(c => {
    const x = Math.round(Math.min(Math.max(0, c + c * luminance), 255)).toString(
      16
    );
    hex += ('00' + x).substr(x.length);
  });
  return hex;
}

function validateHex(color) {
  if (typeof color !== 'string') {
    console.error(`Validate hex color: ${color} is not a string`);
    return false;
  }
  if (color[0] !== '#') {
    console.error(`Validate hex color: ${color} is not a valid hex color`);
    return false;
  }
  color = color.replace('#', '');
  if (color.length < 3) {
    console.error(`Validate hex color: #${color} is not a valid hex color`);
    return false;
  }
  if (color.length === 3) {
    color = color[0] + color[0] + color[1] + color[1] + color[2] + color[2];
  }
  return `#${color}`;
}

function colorHexToRgb(hex) {
  const r = parseInt(hex[1] + hex[2], 16);
  const g = parseInt(hex[3] + hex[4], 16);
  const b = parseInt(hex[5] + hex[6], 16);
  return [r, g, b];
}

export default {
  create,
  styles,
  getContrastYIQ,
  getColorLuminance
};
