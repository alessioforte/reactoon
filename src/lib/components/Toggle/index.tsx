import React, { FC, useState } from 'react';
import { ThemeProvider, withTheme } from 'styled-components';
import { styles } from '../../theme';
import { Props, Data } from './types'
import { Block, Knob, Label } from './styled'

const Toggle: FC<Props> = ({ label, initialValue, name, onChange, theme }) => {
  const [value, setValue] = useState<boolean>(initialValue || false)

  const handleOnClick = (e: any): any => {
    const checked: boolean = !value
    setValue(checked)
    const data: Data = { name, value: checked, checked, label }
    onChange(data, e)
  }

  return (
    <ThemeProvider theme={theme}>
      <Block>
        <Knob
          role='checkbox'
          name={name}
          value={value}
          aria-label={label}
          aria-checked={value}
          onClick={handleOnClick}
          active={value}
          tabIndex={0}
        >
        </Knob>
        {label && <Label>{label}</Label>}
      </Block>
    </ThemeProvider>
  );
}

Toggle.defaultProps = {
  initialValue: false,
  onChange: (value) => console.log(value),
  theme: styles
}

export default withTheme(Toggle);

