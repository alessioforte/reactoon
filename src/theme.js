import { Theme } from './lib';

const theme = Theme.create({
  colors: {
    primary: '#1FA588',
    secondary: '#86C5E0',
    action: '#F7AC91',
    active: '#FFE1C4',
    selected: '#008080',
    hover: '#1FA588',
    success: '#00FA9A',
    warning: '#FFA500',
    danger: '#B22222',
    error: '#B22222',
    info: '#00BFFF',
    light: '#fff',
    dark: '#1a1a1a',
    text: 'red',
    textLight: '#ffffff',
    textDark: '#000000',
    textIdle: '',
    textActive: '',
    textError: '',
    groundzero: '#FFFFFF',
    background: '#F9F9F9',
    flatground: '#EAEAEA',
    ground: '#CCCCCC',
    hillground: '#707070',
    highground: '#303030',
    idle: '#eee',
    disabled: '#778899',
    focus: '#fafafa',
    holiday: '#FF4365',
    shadow: '0px 1px 5px 1px rgba(0, 0, 0, 0.2)'
  },
  borders: {
    radius: 5
  }
});

export default theme;
