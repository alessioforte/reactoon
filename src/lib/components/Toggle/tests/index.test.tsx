import React from 'react'
import { render } from '@testing-library/react'
import Toggle from '../index'
import { Props } from '../interfaces'
import { styles } from '../../../theme'

describe('<Toggle />', () => {
  it('should render and match the snapshot', () => {
    const props: Props = {
      label: 'text',
      onChange: () => {},
      theme: styles

    }
    const { container: { firstChild } } = render(<Toggle {...props}/>)
    expect(firstChild).toMatchSnapshot()
  })
})
