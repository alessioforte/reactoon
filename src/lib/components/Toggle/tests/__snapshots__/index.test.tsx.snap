// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Toggle /> should render and match the snapshot 1`] = `
<div
  class="sc-bdVaJa bSrzPW"
>
  <div
    class="sc-bwzfXH fNnYsN"
  />
  <div
    class="sc-htpNat iULkUd"
  >
    text
  </div>
</div>
`;
